internal/services/jsonrpc/server.go

```diff
@@ -0,0 +1,218 @@
+	// Close the connection when response handling is completed.
+	w.Header().Set("Connection", "close")
+	w.Header().Set("Content-Type", "application/json")
+	r.Close = true
+
+	bodyReader := io.LimitReader(r.Body, szLimit)
+	body, err := io.ReadAll(bodyReader)
+	r.Body.Close()
+	if err != nil {
+		http.Error(w, "error reading request body", http.StatusBadRequest)
+		return
+	}
+	req := new(jsonrpc.Request)
+	err = json.Unmarshal(body, req)
+	if err != nil {
+		resp := jsonrpc.NewErrorResponse(-1, jsonrpc.NewError(jsonrpc.ErrorParse, "invalid request", nil))
+		s.writeJSON(w, resp, http.StatusBadRequest)
+		return
+	}
+	s.processRequest(r.Context(), w, req)
+}
+
+// processRequest handles the jsonrpc.Request with handleRequest to call the
+// appropriate function for the method, creates a response message, and writes
+// it to the http.ResponseWriter.
+func (s *Server) processRequest(ctx context.Context, w http.ResponseWriter, req *jsonrpc.Request) {
+	resp := s.handleRequest(ctx, req)
+	// Some conventions dictate 200 for everything, with the Response.Error
+	// being the only sign of issue. However, a certain set of errors warrant an
+	// http status code.
+	statusCode := http.StatusOK
+	if resp.Error != nil {
+		switch resp.Error.Code {
+		case jsonrpc.ErrorUnknownMethod: // other "not found" is not a 404 since the method at least existed
+			statusCode = http.StatusNotFound // 404
+		case jsonrpc.ErrorInvalidParams, jsonrpc.ErrorInvalidRequest, jsonrpc.ErrorParse:
+			statusCode = http.StatusBadRequest // 400
+		case jsonrpc.ErrorInternal:
+			statusCode = http.StatusInternalServerError // 500
+		}
+	}
+	s.writeJSON(w, resp, statusCode)
+}
+
+// writeJSONWithStatus marshals the provided interface and writes the bytes to
+// the ResponseWriter with the specified response code.
+func (s *Server) writeJSON(w http.ResponseWriter, thing any, code int) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	b, err := json.Marshal(thing)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		s.log.Errorf("JSON encode error: %v", err)
+		return
+	}
+	w.WriteHeader(code)
+	_, err = w.Write(b)
+	if err != nil {
+		s.log.Errorf("Write error: %v", err)
+	}
+}
+
+// type methodHandler func(ctx context.Context, s *Server, params json.RawMessage) (any, *jsonrpc.Error)
+
+// routes maps routes to a handler function.
+//
+// TODO: instead of a map of handlers, create instance of registered request
+// type and call the TxSvc method with it. But I hate reflect.
+// var routes = map[string]methodHandler{
+// 	jsonrpc.MethodAccount:   handleAccount,
+// 	jsonrpc.MethodChainInfo: nil,
+// }
+
+// func registerMethod[T any](method string, req *T, handler func(context.Context, *T) (any, *jsonrpc.Error)) {}
+
+// handleRequest sends the request to the correct handler function if able.
+func (s *Server) handleRequest(ctx context.Context, req *jsonrpc.Request) *jsonrpc.Response {
+	if req.JSONRPC != "2.0" || req.ID == 0 {
```

Comment:
Ah, good catch. This was not smart.  It's only true if the any is an `int(0)` but not an `int64(0)`.  Basically and unset or zero numeric for `ID` is discouraged in requests since in responses a null/unset ID means it's a notification, so the server should reject that malformed request.  Will fix this.